pylot/loggers/eval_metric_logger_operator.py

```diff
@@ -0,0 +1,147 @@
+import erdos
+
+from pylot.utils import LaneMarkingColor, LaneType, LaneMarkingType
+
+
+class EvalMetricLoggerOperator(erdos.Operator):
+    """ Logs the evaluation metrics of the ego-vehicle in the CSV file.
+
+    This operator logs the lateral and longitudinal acceleration and jerk,
+    along with the traffic infractions, lane invasions and collisions
+    committed by the ego vehicle.
+
+    Args:
+        collision_stream (:py:class:`erdos.ReadStream`): Stream on which the
+            collision events are received.
+        lane_invasion_stream (:py:class:`erdos.ReadStream`): Stream on which
+            the lane invasion events are received.
+        traffic_light_invasion_stream (:py:class:`erdos.ReadStream`): Stream
+            on which the traffic light invasion events are received.
+        flags (absl.flags): Object to be used to access the absl flags.
+    """
+
+    def __init__(self, collision_stream, lane_invasion_stream,
+                 traffic_light_invasion_stream, imu_stream, flags):
+        # Save the streams.
+        self._collision_stream = collision_stream
+        self._lane_invasion_stream = lane_invasion_stream
+        self._traffic_light_invasion_stream = traffic_light_invasion_stream
+        self._imu_stream = imu_stream
+
+        # Register callbacks.
+        collision_stream.add_callback(self.on_collision_update)
+        lane_invasion_stream.add_callback(self.on_lane_invasion_update)
+        traffic_light_invasion_stream.add_callback(
+            self.on_traffic_light_invasion_update)
+        imu_stream.add_callback(self.on_imu_update)
+
+        # Initialize a logger.
+        self._flags = flags
+        self._logger = erdos.utils.setup_logging(self.config.name,
+                                                 self.config.log_file_name)
+        self._csv_logger = erdos.utils.setup_csv_logging(
+            self.config.name + '-csv', self.config.csv_log_file_name)
+
+        # Save the last acceleration and timestamp.
+        self._last_lateral_acc, self._last_longitudinal_acc = None, None
+        self._last_timestamp = None
+
+    @staticmethod
+    def connect(collision_stream, lane_invasion_stream,
+                traffic_light_invasion_stream, imu_stream):
+        return []
+
+    def on_collision_update(self, msg):
+        """ Invoked upon receipt of a collision update.
+
+        Logs the collided actor and the intensity of the collision.
+
+        Args:
+            msg (:py:class:`pylot.simulation.messages.CollisionMessage`): The
+                collision message sent by the sensor.
+        """
+        self._csv_logger.info('collision, {}, {}'.format(
```

Comment:
NIT: can you please the white spaces between the commas? I'm not sure if the csv readers automatically remove them. Thus, it may be easier to code plotting scripts if we don't have whitespaces in the first place.

Reply:
Okay, will do!